feat(explorer): collapsible mobile explorer

saberzero1 opened the pull request:
[Screencast from 2024-10-01 23-34-59.webm](https://github.com/user-attachments/assets/a2eee7a4-3b3f-4c88-ade9-910e09a69a0f)


## Progress

- [x] Mobile explorer fully functional
  - [x] Collapsible
  - [x] Icon mode on mobile
- [x] Verify desktop explorer functionality
- [x] Restore styles
  - [x] Desktop layout
  - [x] Full screen mobile layout
- [x] Clean up

saberzero1 committed:
Rewrite mobile explorer

saberzero1 committed:
Finished mobile explorer

saberzero1 committed:
Fix folder/tag page

saberzero1 committed:
Cleanup

saberzero1 committed:
Combine into single explorer tree

saberzero1 committed:
Restore mobile button position

saberzero1 commented:
@aarnphm I suppose this is another PR with a breaking change, as the `Explorer` component is no longer `DesktopOnly`. Keeping it `DesktopOnly` keeps the functionality the same as currently, but makes the mobile `Explorer` view hidden by default.

aarnphm commented on quartz.layout.ts:
```diff
@@ -27,7 +27,7 @@ export const defaultContentPageLayout: PageLayout = {
     Component.MobileOnly(Component.Spacer()),
     Component.Search(),
     Component.Darkmode(),
-    Component.DesktopOnly(Component.Explorer()),
```
We can still keep this i think. we can just say the behaviour in docs or somewhere.

saberzero1 replied:
I have left the `DesktopOnly` out for now for testing. As soon as the code is approved, I'll update the documentation and restore the `DesktopOnly` before the PR is merged.

aarnphm replied:
actually we can just enable this fwiw

aarnphm commented on quartz/components/Explorer.tsx:
```diff
@@ -81,42 +84,70 @@ export default ((userOpts?: Partial<Options>) => {
-            stroke-linejoin="round"
-            class="fold"
+      <div class="explorer-container">
+        <div class={`explorer ${displayClass ?? ""}`}>
```
use `classNames` for consistency

aarnphm commented on quartz/components/Explorer.tsx:
```diff
@@ -81,42 +84,70 @@ export default ((userOpts?: Partial<Options>) => {
+            data-savestate={opts.useSavedState}
+            data-pagepathstate={opts.usePagePath}
+            data-tree={jsonTree}
+            data-mobile={true}
```
add `aria-controls` and `aria-expanded` here for both buttons

aarnphm commented on quartz/components/scripts/explorer.inline.ts:
```diff
@@ -42,75 +41,163 @@ function toggleFolder(evt: MouseEvent) {
+  //prevent scroll under
+  if (document.querySelector("#mobile-explorer")) {
+    const article = document.querySelectorAll(
+      ".popover-hint, footer, .backlinks, .graph, .toc, #progress",
```
this seems, hacky 🤔

saberzero1 replied:
Updated the code to only disable the main content scroll whenever the explorer is open on mobile.

aarnphm commented on quartz/components/scripts/explorer.inline.ts:
```diff
@@ -42,75 +41,163 @@ function toggleFolder(evt: MouseEvent) {
 
 window.addEventListener("resize", setupExplorer)
+
+document.addEventListener("DOMContentLoaded", () => {
```
why are we using DOMContentLoaded here, instead of `nav`?

saberzero1 replied:
I removed the `DOMContentLoaded` listener as the `nav` listener already covers it.

aarnphm reviewed (commented):
cc jacky on this.

saberzero1 commented:
> cc jacky on this.

Thanks for the review.

I think most of your feedback is valid. I'll do some more experimenting to reduce redundant code/hacks.

I have another implementation that is a lot more sophisticated, but requires the navigation tree to be included twice. I remember that this is something we explicitly want to avoid, mostly due to accessibility concerns.

I'll ping you once I am satisfied with my changes.

saberzero1 committed:
Merge branch 'jackyzha0:v4' into explorer

saberzero1 committed:
classNames usage

saberzero1 committed:
Clean up redundant code

saberzero1 committed:
Addressed feedback

saberzero1 committed:
Cleanup

saberzero1 committed:
Restore position of functions

saberzero1 commented:
@aarnphm I believe I have addressed your feedback.

Once the code changes are approved, I'll update the documentation and restore the default configuration (`DesktopOnly` on `Explorer`).

aarnphm commented on quartz/components/Explorer.tsx:
```diff
@@ -81,42 +85,74 @@ export default ((userOpts?: Partial<Options>) => {
+              <polyline points="6 9 12 15 18 9"></polyline>
+            </svg>
+          </button>
+          <div id="explorer-content" class="">
```
why does it have to be an empty class here?

saberzero1 replied:
Fixed.

aarnphm commented on quartz/components/ExplorerNode.tsx:
```diff
@@ -124,9 +125,10 @@ export class FileNode {
    * @returns array containing folder state for tree
    */
-  getFolderPaths(collapsed: boolean): FolderState[] {
+  getFolderPaths(collapsed: boolean, currentFile: string): FolderState[] {
```
doesn't seem like we are using this anywhere in the function?

saberzero1 replied:
Fixed.

aarnphm commented on quartz/components/Footer.tsx:
```diff
@@ -12,7 +13,7 @@ export default ((opts?: Options) => {
     const links = opts?.links ?? []
     return (
-      <footer class={`${displayClass ?? ""}`}>
+      <footer class={classNames(displayClass)}>
```
let's separate this out to a standalone PR?

saberzero1 replied:
Will do.

aarnphm commented on quartz/components/ExplorerNode.tsx:
```diff
@@ -16,6 +16,7 @@ export interface Options {
   folderDefaultState: "collapsed" | "open"
   folderClickBehavior: "collapse" | "link"
   useSavedState: boolean
+  usePagePath: boolean
```
can you explain why we need this?

saberzero1 replied:
`usePagePath`, when enabled, will automatically open the explorer folders to the current page. Similar to how breadcrumbs work. I found this to be nice for navigation between nested pages, especially on mobile. Give it a try!

```ts
Component.Explorer({ usePagePath: true }),
```

aarnphm replied:
seems like a separate feature rather than this one.

saberzero1 replied:
~~Want me to make a separate PR for that?~~

Will move to separate PR.

aarnphm reviewed (approved):
cc jacky on this as well 😄

saberzero1 committed:
Revert Footer

saberzero1 committed:
Addressed feedback

saberzero1 committed:
Remove usePagePath

saberzero1 committed:
Prettier

saberzero1 committed:
Merge branch 'v4' into explorer

aarnphm commented on quartz/components/Footer.tsx:
```diff
@@ -1,6 +1,7 @@
 import { QuartzComponent, QuartzComponentConstructor, QuartzComponentProps } from "./types"
 import style from "./styles/footer.scss"
 import { version } from "../../package.json"
+import { classNames } from "../util/lang"
```
```suggestion
```

saberzero1 committed:
Update quartz/components/Footer.tsx

Co-authored-by: Aaron Pham <Aaronpham0103@gmail.com>

saberzero1 committed:
Merge branch 'v4' into explorer

Stephen-X commented:
Thanks for this!

I'm testing the mobile explorer in my setup. For reasons still unknown to me, the explorer stopped being scrollable in desktop view.

My left sidebar uses a grid layout so that the search bar and the dark mode toggle are on the same row, and the recent notes are pinned to the page bottom.

This is the setting for the explorer in the middle; it has a variable height `max-height: 100%` as I want it to be collapsible. 

```css
.page .sidebar.left > .explorer-container {
    grid-area: 3/1/4/5;
    align-self: flex-start;
    min-height: 3.5rem;
    max-height: 100%;
}
```

<img width="1628" alt="Screenshot 2024-10-13 at 10 11 13 PM" src="https://github.com/user-attachments/assets/66ace470-9a25-456e-bdcc-7f2c89789278">

With this change, unless I set a concrete size to the `max-height` of `explorer-content`, using `max-height: 100%` no longer shows the scrollbar, even though `overflow-y: auto`. I haven't quite figured out yet how to fix this.

saberzero1 commented:
> Thanks for this!
> 
> I'm testing the mobile explorer in my setup. For reasons still unknown to me, the explorer stopped being scrollable in desktop view.
> 
> My left sidebar uses a grid layout so that the search bar and the dark mode toggle are on the same row, and the recent notes are pinned to the page bottom.
> 
> This is the setting for the explorer in the middle; it has a variable height `max-height: 100%` as I want it to be collapsible. 
> 
> ```css
> .page .sidebar.left > .explorer-container {
>     grid-area: 3/1/4/5;
>     align-self: flex-start;
>     min-height: 3.5rem;
>     max-height: 100%;
> }
> ```
> 
> <img width="1628" alt="Screenshot 2024-10-13 at 10 11 13 PM" src="https://github.com/user-attachments/assets/66ace470-9a25-456e-bdcc-7f2c89789278">
> 
> With this change, unless I set a concrete size to the `max-height` of `explorer-content`, using `max-height: 100%` no longer shows the scrollbar, even though `overflow-y: auto`. I haven't quite figured out yet how to fix this.
> 

Can you link the page in question?

Stephen-X commented:
> > ...
> > My left sidebar uses a grid layout so that the search bar and the dark mode toggle are on the same row, and the recent notes are pinned to the page bottom.
> > This is the setting for the explorer in the middle; it has a variable height `max-height: 100%` as I want it to be collapsible.
> > ```css
> > .page .sidebar.left > .explorer-container {
> >     grid-area: 3/1/4/5;
> >     align-self: flex-start;
> >     min-height: 3.5rem;
> >     max-height: 100%;
> > }
> > ```
>> ...
> > With this change, unless I set a concrete size to the `max-height` of `explorer-content`, using `max-height: 100%` no longer shows the scrollbar, even though `overflow-y: auto`. I haven't quite figured out yet how to fix this.
> 
> Can you link the page in question?

Here's a simple repro: https://github.com/Stephen-X/quartz/commit/206f8bc95c8a60fb8ad0942f91c2d16f29ca7e8f. Feel free to pull my branch, it's just serving Quartz doc as test content plus a190cba737f223d8245bd5632bc7a653800dc1be.

I think you merely introduced an additional `explorer-container` wrapper here, so I'm not sure yet what fixed the height that caused `explorer-content` to not be scrollable. Tested several variations, but only setting height to a fixed size worked :(

saberzero1 commented:
> > > ...
> > > My left sidebar uses a grid layout so that the search bar and the dark mode toggle are on the same row, and the recent notes are pinned to the page bottom.
> > > This is the setting for the explorer in the middle; it has a variable height `max-height: 100%` as I want it to be collapsible.
> > > ```css
> > > .page .sidebar.left > .explorer-container {
> > >     grid-area: 3/1/4/5;
> > >     align-self: flex-start;
> > >     min-height: 3.5rem;
> > >     max-height: 100%;
> > > }
> > > ```
> >> ...
> > > With this change, unless I set a concrete size to the `max-height` of `explorer-content`, using `max-height: 100%` no longer shows the scrollbar, even though `overflow-y: auto`. I haven't quite figured out yet how to fix this.
> > 
> > Can you link the page in question?
> 
> Here's a simple repro: https://github.com/Stephen-X/quartz/commit/206f8bc95c8a60fb8ad0942f91c2d16f29ca7e8f. Feel free to pull my branch, it's just serving Quartz doc as test content plus a190cba737f223d8245bd5632bc7a653800dc1be.
> 
> I think you merely introduced an `explorer-container` here, so I'm not sure yet what fixed the height that caused `explorer-content` to not be scrollable. Tested several variations, but only setting height to a fixed size worked :(

I see. That is just a `flex` thing. Any height should work. Try `height: 4px;`.

Stephen-X commented:
> I see. That is just a `flex` thing. Any height should work. Try `height: 4px;`.

Yes, it's nested `flex` element having problem figuring out the parent `flex` element's height.  Unfortunately to make scrolling available, I'll have to set the following line to a fixed size (e.g. `100vh`) rather than 100%, otherwise the mobile explorer stopped being scrollable.

https://github.com/jackyzha0/quartz/blob/741641599b1d90b87e8c8b381ce0f2a5e8fe6775/quartz/components/styles/explorer.scss#L25

Stephen-X commented:
> > I see. That is just a `flex` thing. Any height should work. Try `height: 4px;`.
> 
> Yes, unfortunately to make scrolling available, I'll have to set the following line to a fixed size rather than 100%, otherwise the mobile explorer stopped being scrollable.
> 
> https://github.com/jackyzha0/quartz/blob/741641599b1d90b87e8c8b381ce0f2a5e8fe6775/quartz/components/styles/explorer.scss#L25

Oh I think this is a simple enough thing to fix, it's nested `flex` element having problem figuring out the parent `flex` element's height. Can you change the line above to `height: 100vh`? That should be able to fix this.

saberzero1 commented:
> > > I see. That is just a `flex` thing. Any height should work. Try `height: 4px;`.
> > 
> > Yes, unfortunately to make scrolling available, I'll have to set the following line to a fixed size rather than 100%, otherwise the mobile explorer stopped being scrollable.
> > 
> > https://github.com/jackyzha0/quartz/blob/741641599b1d90b87e8c8b381ce0f2a5e8fe6775/quartz/components/styles/explorer.scss#L25
> 
> Oh I think this is a simple enough thing to fix, it's nested `flex` element having problem figuring out the parent `flex` element's height. I can override the above to `height: 100vh` to fix this; downside is scrollbar is truncated so probably can't ask that you change it in your PR :)

You want me to make a change in order to fix your custom changes?

I'll see if I can clean up the styling, but the explorer should always work on the default Quartz.

If you want to fix your left sidebar without a grid layout, consider something like [this](https://github.com/saberzero1/quartz/blob/1db77f8dfa0829f0c6bdf0dffcbd1288728f9161/quartz/styles/custom.scss#L31-L60)

Stephen-X commented:
> You want me to make a change in order to fix your custom changes?
> 
> I'll see if I can clean up the styling, but the explorer should always work on the default Quartz.
> 
> If you want to fix your left sidebar without a grid layout, consider something like [this](https://github.com/saberzero1/quartz/blob/1db77f8dfa0829f0c6bdf0dffcbd1288728f9161/quartz/styles/custom.scss#L31-L60)

Thanks, but I'm not sure if changing `height` to any thing other than `100%` would be ideal in the official code, that's why I deleted my comment; ideally explorer should take up as much height as possible, folks should override if they want otherwise.

Yes I might go with your `flex`-based setup if this is merged. On second thought this may not be related to `flex` at all. For `overflow` to work, must avoid having nested containers with `height` set to a percentage of the parent container's height which is also a percentage of its parent's.

Thanks for your help! The mobile explorer may be a breaking change for grid-based setup, a major version bump might be needed?

saberzero1 commented:
> > You want me to make a change in order to fix your custom changes?
> > 
> > I'll see if I can clean up the styling, but the explorer should always work on the default Quartz.
> > 
> > If you want to fix your left sidebar without a grid layout, consider something like [this](https://github.com/saberzero1/quartz/blob/1db77f8dfa0829f0c6bdf0dffcbd1288728f9161/quartz/styles/custom.scss#L31-L60)
> 
> Thanks, but I'm not sure if changing `height` to any thing other than `100%` would be ideal in the official code, that's why I deleted my comment; ideally explorer should take up as much height as possible, folks should override if they want otherwise.
> 
> Yes I might go with your `flex`-based setup if this is merged. On second thought this may not be related to `flex` at all. For `overflow` to work, must avoid having nested containers with `height` set to a percentage of the parent container's height which is also a percentage of its parent's.
> 
> Thanks for your help! The mobile explorer may be a breaking change for grid-based setup, a major version bump might be needed?

The sidebars are `display: flex;` by default. I'm not sure if I understand how this would be a breaking change?

Stephen-X commented:
> The sidebars are `display: flex;` by default. I'm not sure if I understand how this would be a breaking change?

Debugging your website, I just realized there isn't an `.explorer-container` wrapper in the solution you committed there: https://github.com/saberzero1/quartz/blob/3a47df05e36c120aaa8f094e4f6c2be728c8b0a2/quartz/components/ExplorerBurger.tsx, maybe that's why scrollbar is showing up?

I think the problem is browsers can't calculate height correctly for `overflow` if there are say 3 or more nested elements with height set as a percentage of their parent's, not sure yet if it's related to them being `flex` containers or not (MDN doesn't state this clearly so I'm not 100% sure if the statement is true: https://developer.mozilla.org/en-US/docs/Web/CSS/overflow#description).

If the solution can keep only the original 2 layers (`.explorer{ #explorer-content{} }`), then all behavior would stay the same way regardless of how forks write their custom layout?

saberzero1 commented:
> > The sidebars are `display: flex;` by default. I'm not sure if I understand how this would be a breaking change?
> 
> Debugging your website, I just realized there isn't an `.explorer-container` wrapper in the solution you committed there: https://github.com/saberzero1/quartz/blob/3a47df05e36c120aaa8f094e4f6c2be728c8b0a2/quartz/components/ExplorerBurger.tsx, maybe that's why scrollbar is showing up?
> 
> I think the problem is browsers can't calculate height correctly for `overflow` if there are say 3 or more nested elements with height set as a percentage of their parent's, not sure yet if it's related to them being `flex` containers or not (MDN doesn't state this clearly so I'm not 100% sure if the statement is true: https://developer.mozilla.org/en-US/docs/Web/CSS/overflow#description).
> 
> If the solution can keep only the original 2 layers (`.explorer{ #explorer-content{} }`), then all behavior would stay the same way regardless of how forks write their custom layout?

There is a reason for the change. It is needed to support mobile exporer without duplicating the entire explorer tree content. My own site is disregarding that requirement by just including it twice.

I'll look into making changes to this PR. I'll keep you posted.

saberzero1 commented:
Alright, I have an idea. The explorer doesn't have to expand on mobile, or can just slide it.

saberzero1 committed:
Changed mobile explorer from dropdown to slide-in

saberzero1 commented:
@Stephen-X Hey, I have a new proposal for the mobile explorer. I want to slide it in from the side instead of dropping down/expanding. See below:

[Screencast from 2024-10-25 18-33-46.webm](https://github.com/user-attachments/assets/a1307404-292a-4fae-b592-3efb0f9a81c4)

We're currently debating between making the top part sticky, so that it stays on top as the user scrolls down, similar to [my own site](https://emilebangma.com/Quartz). In the clip above this is not the case.

If we were to make the top part sticky, it should probably be an option to configure in `quartz.layout.ts` or `quartz.config.ts`, and likely a separate change shortly after this PR is deployed.

Stephen-X commented:
> @Stephen-X Hey, I have a new proposal for the mobile explorer. I want to slide it in from the side instead of dropping down/expanding. See below:
> 
>  Screencast.from.2024-10-25.18-33-46.webm 
> We're currently debating between making the top part sticky, so that it stays on top as the user scrolls down, similar to [my own site](https://emilebangma.com/Quartz). In the clip above this is not the case.
> 
> If we were to make the top part sticky, it should probably be an option to configure in `quartz.layout.ts` or `quartz.config.ts`, and likely a separate change shortly after this PR is deployed.

I think either looks nice (either sliding in from the left or from the top), although in the case of left sliding-in,  it might look better if the hamburger menu rotates counterclockwise to go with the slide-in animation like a knob 😄

I personally made the left sidebar `sticky` as well in mobile mode, wrote a plugin to auto-hide on scrolling 😆 Can contribute a PR if there's interest for a mobile title bar in the main repo. Here's my website: https://stepht.org/ 

---

> > If the solution can keep only the original 2 layers (`.explorer{ #explorer-content{} }`), then all behavior would stay the same way regardless of how forks write their custom layout?
> 
> There is a reason for the change. It is needed to support mobile exporer without duplicating the entire explorer tree content. My own site is disregarding that requirement by just including it twice.
>
> I'll look into making changes to this PR. I'll keep you posted.

Regarding the reported issue with the extra `explorer-container` making the file list non-scrollable when the container itself has a `max-height: 100%` set on it in custom layout (as was also the case in your personal website setup last time I checked), I just did a quick test, I'm able to remove the extra `explorer-container` without much of an issue. Here's a sample commit (please note that the css part is crude, I mostly just moved everything out of `.explorer-container`; might need to move some stuff into `.explorer`): https://github.com/Stephen-X/quartz/commit/76187d6ea7189eac07fa203bdac8b00d92e1f591.

And now the file list is scrollable again in desktop mode, confirming my theory before about `overflow`:

<img width="449" alt="Screenshot 2024-10-25 at 11 01 52 PM" src="https://github.com/user-attachments/assets/22461c38-ac6e-41d5-9443-4f166394afcd">

Here's what it looks like in mobile mode, I made some small adjustments to styling to make the icon vertically centered again after `explorer-container` is removed, see my notes in `explorer.scss`:

<img width="449" alt="Screenshot 2024-10-25 at 11 01 06 PM" src="https://github.com/user-attachments/assets/cf8f16a9-e68f-4eaa-b5cf-44af597d6fef">

I'm not entirely sure what you were referring to regarding "duplicating the entire explorer tree content" if the container is removed. Please feel free to refer to the commit to see if there's anything breaking with the change.

With `explorer-container` removed the mobile explorer feature becomes a (mostly) non-breaking change to fork's existing custom layouts or styles on `.explorer`.

Btw, I also changed `display: flex` to `display:inline-block` for `.explorer` / `.explorer-container`, `flex` is shrinking the hamburger menu icon when there isn't enough space in the left sidebar, making it harder to tap on; also that's different behavior from other components.

saberzero1 committed:
Remove container div

saberzero1 committed:
Restore position

saberzero1 committed:
Slide in + no ugly onload animation on mobile

saberzero1 commented:
[Screencast from 2024-10-26 17-58-29.webm](https://github.com/user-attachments/assets/eedb0f33-7489-461b-9f19-f78863488864)

saberzero1 committed:
Cleanup

Stephen-X commented on quartz/components/styles/explorer.scss:
```diff
@@ -1,14 +1,69 @@
+}
+
 .explorer {
   display: flex;
```
Might be better showing explorer as `inline-block`.

As reported, flex shrinks if there isn't enough space. So if the title bar becomes squeezy the explorer icon will become smaller, making it harder to tap on and also impacting `explorer-content` alignment, as demoed below:

## Before

<img width="437" alt="Screenshot 2024-11-02 at 2 26 46 AM" src="https://github.com/user-attachments/assets/b07eb367-c25a-47a3-bd55-d5a4a1b1e603">

## After

<img width="437" alt="Screenshot 2024-11-02 at 2 36 51 AM" src="https://github.com/user-attachments/assets/89a00cef-3302-4c5c-add5-10fa61679f40">

saberzero1 replied:
Can you verify this is not related to your use of a customized grid layout for your left sidebar over Quartz's default flex for the left sidebar?

Stephen-X replied:
I didn't alter the flex layout in mobile mode.

<img width="502" alt="Screenshot 2024-11-02 at 3 00 05 AM" src="https://github.com/user-attachments/assets/4b05a741-41fa-484f-86da-ea7efa795d19">


I think this is easily reproducible if you make the title bar more squeezy by having a longer page title or use "responsive" dimension in Chromium browsers and make the page narrower. It's really related to the `explorer` container itself being a `flex` container causing its containing component (the hamburger button) to shrink.

<img width="453" alt="Screenshot 2024-11-02 at 3 01 37 AM" src="https://github.com/user-attachments/assets/372c24db-2680-411f-81f3-15c89b331cb2">

saberzero1 replied:
Thanks. I'll look into it.

I asked because your previous suggested change didn't work in default Quartz.

I'll let you know my conclusion soon.

Once again, thanks for the reports!

Stephen-X commented:
Bug report: Because explorer is hidden in folder index pages, the mobile explorer button also disappears from title bar.

saberzero1 commented:
> Bug report: Because explorer is hidden in folder index pages, the mobile explorer button also disappears from title bar.

Can you check if this stil l occurs when the explorer is added to listpages in `quartz.layout.ts`?

Stephen-X commented:
> > Bug report: Because explorer is hidden in folder index pages, the mobile explorer button also disappears from title bar.
> 
> Can you check if this stil l occurs when the explorer is added to listpages in `quartz.layout.ts`?

Sorry that was a careless bug report. My setup mistake while testing the PR in different cases. I forgot I made explorer `desktop-only` for ListPages before this PR because the whole explorer would otherwise appear in the title bar.

saberzero1 committed:
Icon spacing

PatMulligan commented:
Big fan of this PR 📣

saberzero1 committed:
Merge branch 'v4' into explorer

saberzero1 committed:
SASS styling fixes

saberzero1 committed:
Remove unused parameter

saberzero1 committed:
Copilot nitpick

saberzero1 committed:
Prefer compatibility

saberzero1 commented:
@aarnphm this should be ready now.

aarnphm commented on quartz/components/styles/explorer.scss:
```diff
@@ -1,14 +1,70 @@
+@media all and ($mobile) {
+  .page > #quartz-body {
+    // Shift page position when toggling Explorer on mobile.
+    & > :not(.sidebar.left:has(.explorer)) {
```
Ig we will always leave explorer to the left sidebar? If users move explorer to the right sidebar this wouldn't work

saberzero1 replied:
The problem is that the right sidebar on mobile is at the very bottom of the page.

aarnphm replied:
ah i see, makes sense then

aarnphm reviewed (approved):
I have one comments,otherwise LGTM
